File: 463_island_perimeter.java
class Solution {
    public int islandPerimeter(int[][] grid) {
        int perimeter = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    perimeter += countNumSidesUntouched(grid, i, j);
                }
            }
        }

        return perimeter;
    }

    private int countNumSidesUntouched(int[][] grid, int x, int y) {
        int rows = grid.length, cols = grid[0].length;
        int remaining = 4;
        if (x - 1 >= 0) {
            remaining -= (grid[x - 1][y] == 1) ? 1 : 0;
        }
        if (y - 1 >= 0) {
            remaining -= (grid[x][y - 1] == 1) ? 1 : 0;
        }
        if (x + 1 < rows) {
            remaining -= (grid[x + 1][y] == 1) ? 1 : 0;
        }
        if (y + 1 < cols) {
            remaining -= (grid[x][y + 1] == 1) ? 1 : 0;
        }
        return remaining;
    }
}

File: python/463_island_perimeter.py
class Solution:
    def islandPerimeter_substract(self, grid: List[List[int]]) -> int:
        # calculate number of 1 to get perimeter
        perimeter = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    perimeter += 4
                    # check adjacent block and remove overlapping edge
                    for dx, dy in [[0, -1], [1, 0], [0, 1], [-1, 0]]:
                        x, y = i + dx, j + dy
                        if x >= 0 and x < len(grid) and y >= 0 and y < len(grid[0]) and grid[x][y] == 1:
                            perimeter -= 1
        return perimeter

    def islandPerimeter(self, grid: List[List[int]]) -> int:
        # calculate number of adjacent 0 to get perimeter
        perimeter = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    for dx, dy in [[0, -1], [1, 0], [0, 1], [-1, 0]]:
                        x, y = i + dx, j + dy
                        if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[0]) or grid[x][y] == 0:
                            perimeter += 1
        return perimeter
